pkg/helm/jsonnet_test.go

```diff
@@ -140,3 +137,121 @@ func TestIncludeCrdsFalse(t *testing.T) {
+		KubeVersion: kubeVersion,
+		IncludeCRDs: true,
+	}
+	inputOptionsFromJsonnet := make(map[string]interface{})
+	inputOptionsFromJsonnet["values"] = values
+	inputOptionsFromJsonnet["calledFrom"] = calledFrom
+	inputOptionsFromJsonnet["kubeVersion"] = kubeVersion
+
+	helmMock := &MockHelm{}
+	// ChartExists called on both function calls, only template is cached
+	helmMock.On(
+		"ChartExists",
+		"exampleChartPath",
+		mock.AnythingOfType("*helm.JsonnetOpts")).
+		Return("/full/chart/path", nil).
+		Twice()
+	// this verifies that the helmMock.Template() method is called with the
+	// correct arguments and only a single time.
+	helmMock.On("Template", "exampleChartName", "/full/chart/path", expectedHelmTemplateOptions).
+		Return(manifest.List{}, nil).
+		Once()
+
+	nf := NativeFunc(helmMock)
+	require.NotNil(t, nf)
+
+	// the mandatory parameters to helm.template() in Jsonnet
+	params := []string{
+		"exampleChartName",
+		"exampleChartPath",
+	}
+
+	// mandatory parameters + the k-v pairs from the Jsonnet input
+	paramsInterface := make([]interface{}, 3)
+	paramsInterface[0] = params[0]
+	paramsInterface[1] = params[1]
+	paramsInterface[2] = inputOptionsFromJsonnet
+
+	_, err := nf.Func(paramsInterface)
+	firstCommandArgs := helmMock.TestData().Get("templateCommandArgs").StringSlice()
+	require.NoError(t, err)
+	_, err = nf.Func(paramsInterface)
+	secondCommandArgs := helmMock.TestData().Get("templateCommandArgs").StringSlice()
+	require.NoError(t, err)
+
+	helmMock.AssertExpectations(t)
+
+	// Verify command line args are same between the two calls
+	require.Equal(t, firstCommandArgs, secondCommandArgs)
+}
+
+type templateData struct {
+	chartName string
+	chartPath string
+	opts      TemplateOpts
+}
+
+var templateTestCases = []struct {
+	name       string
+	data       templateData
+	errMessage string
+}{
+	{
+		name: "emptyData",
+		data: templateData{
+			chartName: "testChart",
+			chartPath: "./chart/path",
+			opts:      TemplateOpts{},
+		},
+	},
+	{
+		name: "fullData",
+		data: templateData{
+			chartName: "bigChart",
+			chartPath: "./chart/bigPath",
+			opts: TemplateOpts{
+				map[string]interface{}{
+					"installCRDs": true,
+					"multitenancy": map[string]interface{}{
+						"enabled":               false,
+						"defaultServiceAccount": "default",
+						"privileged":            false,
+					},
+					"clusterDomain": "cluster.local",
+					"cli": map[string]interface{}{
+						"image":        "test-image.io",
+						"nodeSelector": map[string]interface{}{},
+						"tolerations":  []interface{}{},
+					},
+					"baz": []int32{12, 13},
+				},
+				[]string{"asdf", "qwer", "zxcv"},
+				true,
+				false,
+				"version",
+				"namespace",
+				false,
+			},
+		},
+	},
+}
+
+func BenchmarkTemplateKey(b *testing.B) {
+	for _, c := range templateTestCases {
+		b.Run(c.name, func(b *testing.B) {
+			for i := 0; i < b.N; i++ {
+				// nolint:errcheck
+				templateKey(c.data.chartName, c.data.chartPath, c.data.opts)
```

Comment:
Interesting, turns out that json.Marshal really is the fastest option (of the few I tested)